crypto/bio/bf_lbuf.c

```diff
@@ -264,16 +294,17 @@ static long linebuffer_ctrl(BIO *b, int cmd, long num, void *ptr)
 
         for (;;) {
             BIO_clear_retry_flags(b);
-            if (ctx->obuf_len > 0) {
+            if (ctx->obuf_len > ctx->prefix_len) {
                 r = BIO_write(b->next_bio, ctx->obuf, ctx->obuf_len);
                 BIO_copy_next_retry(b);
                 if (r <= 0)
                     return (long)r;
-                if (r < ctx->obuf_len)
-                    memmove(ctx->obuf, ctx->obuf + r, ctx->obuf_len - r);
-                ctx->obuf_len -= r;
+                if (r < ctx->obuf_len && r > ctx->prefix_len)
+                    memmove(ctx->obuf + ctx->prefix_len, ctx->obuf + r,
+                            ctx->obuf_len - r - ctx->prefix_len);
+                ctx->obuf_len -= r - ctx->prefix_len;
```

Comment:
Hmmmm....this is interesting. If `obuf` contains "PREFIX my string to write out\n", and the `BIO_write()` call returns (say)  12, then after all of the above `obuf` will now contain "PREFIXring to write out\n". Doesn't that mean in the next flush we get the PREFIX appearing in the middle of a line? Or am I misunderstanding this? i.e. what is actually written is "PREFIX my stPREFIXring to write out\n"

Reply:
Good point!

The way I've been thinking about this filter is that this should never happen...  that whole line should be written out in one go!  Of course, this wasn't a problem before, but is now.  I'll give this a think.